### ISSUE TITLE
How to deal with provider specific configuration

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#1567]
We've recently introduced a `runtime` property which is available in the `serverless.yaml` file.
This is needed for AWS deployments as AWS needs to know what runtime the function should use.

However it seems like Azure, Google, IBM etc. don't necessarily have those concepts (at least as far as I've looked into them so far) so it's kind of an "only used by AWS" feature.

Should we keep it (it's also a kind of documentation to show devs what runtime the service is written in even if it's not used by the provider).

What do you think @eahefnawy @flomotlik @ac360 ?

**Comment by flomotlik** [comment]
Imho if its not necessary we shouldn't force them to do it. Another interesting question here is how we deal with provider specific configuration that should be set for the whole service, e.g. Runtime, Memory, or any other provider specific config you want to set.

Imho 2 options

1. We leave those general config items at the root of the serverless.yml. Makes it less structured and maybe a bit easier to read at first
2. We create a `config` property that has things like Runtime, Memory, or any other provider specific config that should be set. Then each plugin could potentially define required attributes that have to be set otherwise the system throws an error right away.

Even though it makes the config file a little bit more complex I think I'm in favour of the second approach including integrated validation for it. Any thoughts on that @pmuens @eahefnawy @ac360

**Comment by eahefnawy** [comment]
I've thought about that before and my conclusion is that it's just about where we validate each property in `serverless.yaml`. If we know its a common property between providers, then the validation happens in the Service class during loading, if it's only used by few/single provider, the validation should happen in the relevant provider plugin.

Take the `handler` property for example, it's required by aws only, so we validate it in the aws plugins, another deeper example is the APIG `method` property, we only validate it if we actually have an `http` property defined. I think it could be just as simple with the `runtime` property.

**Comment by pmuens** [comment]
IMO the "basic" / "general" `serverless.yaml` should only include the necessary config which is used by all providers (the stuff where we have an overlap between all providers). Otherwise the `serverless.yaml` file gets more and more bloated.

We could wrap provider specific stuff inside a provider property such as this:

```
aws:
  runtime: nodejs4.3
```

We may even use this property to check which provider is used (although this means that each provider needs a configuration).

**Comment by flomotlik** [comment]
I would keep the configuration key the same across providers, just let them embed different values. e.g.

```yaml
name: my_service
provider: aws
config:
  runtime: nodejs4.3
  memory: 1024
```

```yaml
name: my_service
provider: azure
config:
  key1: DontKnowAnyAzureConfigs
  key2: StillDontknowAny
```

This way we have a clear and unambiguous configuration for each provider with the `provider` and `config` keys. when using different providers people will always know where to put things, otherwise they might have to think is it `azure` or is it `microsoft`, is it `ibm` or is it `openwhisk` for the config property.

Then each plugin can set required attributes (those could be set only for config or imho in general). We don't have a concept of *general config* that spans across providers because we might add another provider in the future that doesn't support this *general config*. Each config simply defines its required attributes and help can also print those for example and validate all are there.

**Comment by flomotlik** [comment]
And I agree with @pmuens that the highest level of the serverless.yml should only contain elements that are used across providers.

**Comment by eahefnawy** [comment]
I think we're over engineering this... @pmuens which step/plugin exactly needs this `runtime` property? cant you just validate it there?

**Comment by eahefnawy** [comment]
right now the highest level of the file has the following: `name` `provider` `functions` and `resources` which are all used by all providers and validated in the Service class...the `default` top level property is not validated/required cause it's optional, we can do the same with `runtime` but just validate it right where we need it.

**Comment by pmuens** [comment]
I like the `config` property as a single place to store provider related stuff.

I would not put provider related config setup inside the root of the `serverless.yaml` file because wrapping it inside the `config` object makes it more reasonable that it's configuration and we can communicate this to Plugin authors that that's the place where they're about to get provider related configuration.

Furthermore it's easier for the user to see a pattern when he switches between templates.

**Comment by flomotlik** [comment]
agree with @pmuens . Imho if we start allowing things like runtime in the root we'll have to allow a ton of things in root without any control over it in the future and it will inevitably end in a mess.

This tweet is always a good reminder for me in those situations: https://twitter.com/solomonstre/status/715361557408931840

**Comment by ac360** [comment]
Hmm.   Don't we have `defaults` already?  I thought `defaults` does what you guys suggest`config` should do?

Aren't these the only steps necessary:

* Make sure `serverless.yml` validation is handled by provider-specific plugins.
* If `defaults` exists, the provider-specific plugin should check if its subproperties are valid.
* If `defaults` does not exist, the provider-specific plugin should check if the **function** subproperties are valid.

**Comment by pmuens** [comment]
Yes. We have the default implementation. Although I would rename it from `default` to `config` or something else as `default` is less expressive IMHO.

**Comment by flomotlik** [comment]
Yup agree with @pmuens that config is imho clearer than default

**Comment by ac360** [comment]
Hmm, but the whole file is config, which makes the property seem a bit vague.

Looking at @eahefnawy's new API Gateway issues, this vagueness is clear because his instincts are to put other data (outside of functions) in there.  However, his idea has a lot of merits.  A `config` object to set values for common properties, whether they be function config or api gateway config (e.g., caching) seems pretty handy.

**Comment by flomotlik** [comment]
> Hmm, but the whole file is config, which makes the property seem a bit vague.

Yup that makes total sense. Not clear what its for if its just config, or how it relates to provider or other things.

Imho 2 ways forward (if you have others please propose):

1. Make the provider config optionally an object

```
provider: aws
```

```
provider:
  name: aws
  runtime: nodejs
  ...
```

2. Special provider config field (or something similarly named)

```
provider: aws
provider_config:
  memory: 1024
runtime: nodejs
```

The main issue I see with both is that its pretty provider focused, but I think I prefer the first option with making the provider an object

**Comment by ac360** [comment]
@flotmotlik Great thinking w/ the first option.  I'm going to keep pondering it.  But ATM, I like option #1 because it means less root properties!  Check out @eahefnawy's API Gateway issues.  Function config may need to be put in another sub-property, to make room for other useful data in `provider`.

**Comment by eahefnawy** [comment]
@ac360 yep we still got serverless variables that you can reference. I'm with ya about the config key, but I still not feeling comfortable putting the config at the highest level for the reasons @flomotlik mentioned before. I like @flomotlik ideas above by setting a provider object. it's a patter that we are already following

**Comment by pmuens** [comment]
Also like the idea to make it optionally an object and keep everything provider related there.

However I also think that this should be a place for more general provider related configs and not event specific configuration such as APIG. This config should live inside the events object of the corresponding event.

**Comment by eahefnawy** [comment]
@pmuens what if a certain event config is not tied to a specific function? The `events` object is inside each function, so we can't use that. We gotta figure out a way to set the "general" event config that is applicable to the entire service (rather than the function).

IMO if it's a service wide config (whether event related or not), it should live inside that `provider` object, otherwise there's nowhere else to put it.

**Comment by pmuens** [comment]
@eahefnawy yes. Good catch! That might be a common problem.

The only thing I would like is to separate the event config as much as possible from the "general" provider config (and vice versa) so that it's only added there when there's no other way around it. 

But if this is not possible we should move it somewhere else. The provider config is fine for me.

**Comment by flomotlik** [comment]
Implemented in #1635
